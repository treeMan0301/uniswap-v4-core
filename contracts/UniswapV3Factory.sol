// SPDX-License-Identifier: UNLICENSED
pragma solidity =0.7.6;

import './interfaces/IUniswapV3Factory.sol';

import './UniswapV3Pair.sol';
import './UniswapV3PairDeployer.sol';

contract UniswapV3Factory is IUniswapV3Factory, UniswapV3PairDeployer {
    address public override owner;

    mapping(uint24 => int24) public override feeAmountTickSpacing;
    uint24[] public override allEnabledFeeAmounts;

    mapping(address => mapping(address => mapping(uint24 => address))) public override getPair;

    function allEnabledFeeAmountsLength() external view override returns (uint256) {
        return allEnabledFeeAmounts.length;
    }

    constructor() {
        owner = msg.sender;
        emit OwnerChanged(address(0), msg.sender);

        enableFeeAmount(600, 12);
        enableFeeAmount(3000, 60);
        enableFeeAmount(9000, 180);
    }

    function createPair(
        address tokenA,
        address tokenB,
        uint24 fee
    ) external override returns (address pair) {
        require(tokenA != tokenB, 'A=B');
        (address token0, address token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
        require(token0 != address(0), 'A=0');
        int24 tickSpacing = feeAmountTickSpacing[fee];
        require(tickSpacing != 0, 'FNA');
        require(getPair[token0][token1][fee] == address(0), 'PAE');
        pair = deploy(address(this), token0, token1, fee, tickSpacing);
        getPair[token0][token1][fee] = pair;
        // populate mapping in the reverse direction, deliberate choice to avoid the cost of comparing addresses
        getPair[token1][token0][fee] = pair;
        emit PairCreated(token0, token1, fee, tickSpacing, pair);
    }

    function setOwner(address _owner) external override {
        require(msg.sender == owner, 'OO');
        emit OwnerChanged(owner, _owner);
        owner = _owner;
    }

    function enableFeeAmount(uint24 fee, int24 tickSpacing) public override {
        require(msg.sender == owner, 'OO');
        require(fee < 1000000, 'FEE');
        require(tickSpacing > 0, 'TS');
        require(feeAmountTickSpacing[fee] == 0, 'FAI');

        feeAmountTickSpacing[fee] = tickSpacing;
        allEnabledFeeAmounts.push(fee);
        emit FeeAmountEnabled(fee, tickSpacing);
    }
}
