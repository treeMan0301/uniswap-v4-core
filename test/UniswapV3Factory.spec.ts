import {ethers, waffle} from 'hardhat'
import {UniswapV3Factory} from '../typechain/UniswapV3Factory'
import {expect} from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'

import {FeeAmount, getCreate2Address, TICK_SPACINGS} from './shared/utilities'

const TEST_ADDRESSES: [string, string] = [
  '0x1000000000000000000000000000000000000000',
  '0x2000000000000000000000000000000000000000',
]

describe('UniswapV3Factory', () => {
  const [wallet, other] = waffle.provider.getWallets()

  let factory: UniswapV3Factory
  let pairBytecode: string
  beforeEach('deploy factory', async () => {
    const factoryFactory = await ethers.getContractFactory('UniswapV3Factory')
    pairBytecode = (await ethers.getContractFactory('UniswapV3Pair')).bytecode
    factory = (await factoryFactory.deploy(await wallet.getAddress())) as UniswapV3Factory
  })

  it('initial owner is deployer', async () => {
    expect(await factory.owner()).to.eq(wallet.address)
  })

  it('initial pairs length is 0', async () => {
    expect(await factory.allPairsLength()).to.eq(0)
  })

  async function createAndCheckPair(tokens: [string, string], feeAmount: FeeAmount) {
    const create2Address = getCreate2Address(factory.address, tokens, feeAmount, TICK_SPACINGS[feeAmount], pairBytecode)
    const create = factory.createPair(tokens[0], tokens[1], feeAmount)

    await expect(create)
      .to.emit(factory, 'PairCreated')
      .withArgs(TEST_ADDRESSES[0], TEST_ADDRESSES[1], feeAmount, TICK_SPACINGS[feeAmount], create2Address, 1)

    await expect(factory.createPair(tokens[0], tokens[1], feeAmount)).to.be.revertedWith(
      'UniswapV3Factory::createPair: pair already exists'
    )
    await expect(factory.createPair(tokens[1], tokens[0], feeAmount)).to.be.revertedWith(
      'UniswapV3Factory::createPair: pair already exists'
    )
    expect(await factory.getPair(tokens[0], tokens[1], feeAmount), 'getPair in order').to.eq(create2Address)
    expect(await factory.getPair(tokens[1], tokens[0], feeAmount), 'getPair in reverse').to.eq(create2Address)
    expect(await factory.allPairs(0), 'first pair in allPairs').to.eq(create2Address)
    expect(await factory.allPairsLength(), 'number of pairs').to.eq(1)

    const pairContractFactory = await ethers.getContractFactory('UniswapV3Pair')
    const pair = pairContractFactory.attach(create2Address)
    expect(await pair.factory(), 'pair factory address').to.eq(factory.address)
    expect(await pair.token0(), 'pair token0').to.eq(TEST_ADDRESSES[0])
    expect(await pair.token1(), 'pair token1').to.eq(TEST_ADDRESSES[1])
    expect(await pair.fee(), 'pair fee').to.eq(feeAmount)
    expect(await pair.tickSpacing(), 'pair tick spacing').to.eq(TICK_SPACINGS[feeAmount])
  }

  describe('#createPair', () => {
    it('succeeds for low fee pair', async () => {
      await createAndCheckPair(TEST_ADDRESSES, FeeAmount.LOW)
    })

    it('succeeds for medium fee pair', async () => {
      await createAndCheckPair(TEST_ADDRESSES, FeeAmount.MEDIUM)
    })

    it('succeeds for high fee pair', async () => {
      await createAndCheckPair(TEST_ADDRESSES, FeeAmount.HIGH)
    })

    it('succeeds if tokens are passed in reverse', async () => {
      await createAndCheckPair([TEST_ADDRESSES[1], TEST_ADDRESSES[0]], FeeAmount.MEDIUM)
    })

    it('gas', async () => {
      await snapshotGasCost(factory.createPair(TEST_ADDRESSES[0], TEST_ADDRESSES[1], FeeAmount.MEDIUM))
    })
  })

  describe('#setOwner', () => {
    it('fails if caller is not owner', async () => {
      await expect(factory.connect(other).setOwner(wallet.address)).to.be.revertedWith(
        'UniswapV3Factory::setOwner: must be called by owner'
      )
    })

    it('updates owner', async () => {
      await factory.setOwner(other.address)
      expect(await factory.owner()).to.eq(other.address)
    })

    it('cannot be called by original owner', async () => {
      await factory.setOwner(other.address)
      await expect(factory.setOwner(wallet.address)).to.be.revertedWith(
        'UniswapV3Factory::setOwner: must be called by owner'
      )
    })
  })
})
