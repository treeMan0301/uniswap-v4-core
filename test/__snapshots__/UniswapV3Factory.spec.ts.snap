// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniswapV3Factory #createPair gas 1`] = `4158398`;
